File: xchange-idexjaxrs/src/main/java/org/knowm/xchange/idex/IdexAccountService.java
package org.knowm.xchange.idex;


import kotlin.NotImplementedError;
import kotlin.jvm.internal.Intrinsics;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.knowm.xchange.currency.Currency;
import org.knowm.xchange.dto.account.AccountInfo;
import org.knowm.xchange.dto.account.Balance;
import org.knowm.xchange.dto.account.FundingRecord;
import org.knowm.xchange.dto.account.FundingRecord.Status;
import org.knowm.xchange.dto.account.FundingRecord.Type;
import org.knowm.xchange.dto.account.Wallet;
import org.knowm.xchange.idexjaxrs.dto.*;
import org.knowm.xchange.idexjaxrs.service.ReturnCompleteBalancesApi;
import org.knowm.xchange.idexjaxrs.service.ReturnDepositsWithdrawalsApi;
import org.knowm.xchange.idexjaxrs.service.WithdrawApi;
import org.knowm.xchange.service.account.AccountService;
import org.knowm.xchange.service.trade.params.TradeHistoryParams;
import org.knowm.xchange.service.trade.params.WithdrawFundsParams;
import si.mazi.rescu.RestProxyFactory;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

import static org.knowm.xchange.idex.IdexExchange.Companion.safeParse;


public class IdexAccountService implements AccountService {
    private final String apiKey;
    private final IdexExchange idexExchange;

    public IdexAccountService(@NotNull IdexExchange idexExchange) {
        Intrinsics.checkParameterIsNotNull(idexExchange, "idexExchange");
        this.idexExchange = idexExchange;
        apiKey = this.idexExchange.getExchangeSpecification().getApiKey();
    }



    public final String getApiKey() {
        return apiKey;
    }

    @NotNull
    public AccountInfo getAccountInfo() {
        AccountInfo ret = null;
        try {
            String apiKey = idexExchange.getExchangeSpecification().getApiKey();
            String s = apiKey.substring(0, 6) + "…";
            ReturnCompleteBalancesApi proxy = RestProxyFactory.createProxy(ReturnCompleteBalancesApi.class, idexExchange.getExchangeSpecification().getSslUri());
            ReturnCompleteBalancesResponse returnBalancesPost = null;
            ret = null;
            returnBalancesPost = proxy.completeBalances(new CompleteBalancesReq().address(apiKey));

            ret = new AccountInfo(new Wallet(s, returnBalancesPost.entrySet().stream().map(entry -> {
                Balance balance = new Balance(new Currency(entry.getKey()), null, entry.getValue().getAvailable(), entry.getValue().getOnOrders());
                return balance;
            }).collect(Collectors.toList())));


        } catch (Exception e) {


        }
        return ret;
    }


    @NotNull
    public String requestDepositAddress(@Nullable org.knowm.xchange.currency.Currency currency, @NotNull String... args) {
        return idexExchange.getExchangeSpecification().getApiKey();
    }

    @Override
    @NotNull
    public List<FundingRecord> getFundingHistory(@NotNull TradeHistoryParams params) {
        List<FundingRecord> ret = null;
        if (!(params instanceof IdexTradeHistoryParams)) {
            throw new Error("getFundingHistory requires " + IdexTradeHistoryParams.class.getCanonicalName());
        } else {
            try {
                ret= mutableList(RestProxyFactory.createProxy(ReturnDepositsWithdrawalsApi.class, idexExchange.getDefaultExchangeSpecification().getSslUri()).fundingHistory((DepositsWithdrawalsReq) params));
            } catch (Exception e) {


            }
        }
        return ret;
    }

    private final List<FundingRecord> mutableList(ReturnDepositsWithdrawalsResponse returnDepositsWithdrawalsPost) {
        return (List<FundingRecord>) Arrays.asList(returnDepositsWithdrawalsPost.getWithdrawals().stream().map(
                fundingLedger -> new FundingRecord(
                        apiKey,
                        new Date(Long.parseLong(fundingLedger.getTimestamp()) * 1000),
                        new Currency(fundingLedger.getCurrency()),
                        safeParse(fundingLedger.getAmount()),
                        fundingLedger.getTransactionHash(),
                        fundingLedger.getDepositNumber(),
                        Type.WITHDRAWAL,
                        Status.resolveStatus(fundingLedger.getStatus()),
                        BigDecimal.ZERO,
                        BigDecimal.ZERO,
                        ""
                )).collect(Collectors.toList()),
                returnDepositsWithdrawalsPost.getDeposits().stream().map(
                        fundingLedger1 -> new FundingRecord(
                                apiKey,
                                new Date(Long.parseLong(fundingLedger1.getTimestamp()) * 1000),
                                new Currency(fundingLedger1.getCurrency()),
                                safeParse(fundingLedger1.getAmount()),
                                fundingLedger1.getTransactionHash(),
                                fundingLedger1.getDepositNumber(),
                                Type.WITHDRAWAL,
                                Status.resolveStatus(fundingLedger1.getStatus()),
                                BigDecimal.ZERO,
                                BigDecimal.ZERO,
                                ""
                        )).collect(Collectors.toList()))
                        .stream().flatMap(List::stream)
                        .sorted(Comparator.comparing(FundingRecord::getDate));
    }


    @NotNull
    public TradeHistoryParams createFundingHistoryParams() {
        return new IdexTradeHistoryParams();
    }

    @NotNull
    public String withdrawFunds(@NotNull WithdrawFundsParams w) {
         String ret="error";
        if (w instanceof IdexWithdraw) {
            try {
                WithdrawResponse withdraw = RestProxyFactory.createProxy(WithdrawApi.class, idexExchange.getDefaultExchangeSpecification().getSslUri()).withdraw((WithdrawReq) w);
                ret = withdraw.toString() ;
            } catch (Exception e) {


            }

        } else {
            throw new Error("withdraw method requires " + org.knowm.xchange.idex.IdexWithdraw.class.getCanonicalName());
        }
        return ret;
    }

    @Override
    @NotNull
    public String withdrawFunds(@NotNull Currency currency, @NotNull BigDecimal amount, @NotNull String address) {
//todo: validate the RSV from the WithdrawlReq created
        //        Intrinsics.checkParameterIsNotNull(currency, "currency");
//        Intrinsics.checkParameterIsNotNull(amount, "amount");
//        Intrinsics.checkParameterIsNotNull(address, "address");
//        new IdexWithdraw;
//        String var10003 = this.apiKey;
//        Intrinsics.checkExpressionValueIsNotNull(this.apiKey, "apiKey");
//        Intrinsics.checkExpressionValueIsNotNull(amount.toString(), "amount.toString()");
//        Intrinsics.checkExpressionValueIsNotNull(currency.getSymbol(), "currency!!.symbol");
//        Intrinsics.checkExpressionValueIsNotNull(this.idexExchange.getTradeService().getIdexServerNonce(), "idexExchange.tradeService.idexServerNonce");
//        String var4 = "provide threadlocals? for S";
        throw  (new NotImplementedError("An operation is not implemented: withdraw signature verification " /*+ var4*/));
    }


}

File: xchange-idexjaxrs/src/main/kotlin/org/knowm/xchange/idexjaxrs/IdexAccountService.kt
package org.knowm.xchange.idexjaxrs

import org.knowm.xchange.currency.Currency
import org.knowm.xchange.dto.account.*
import org.knowm.xchange.idexjaxrs.dto.*
import org.knowm.xchange.idexjaxrs.service.*

import org.knowm.xchange.service.account.*
import org.knowm.xchange.service.trade.params.*
import java.math.*
import java.util.*

class IdexAccountService(private val idexjaxrsExchange: IdexjaxrsExchange) : AccountApi(),
        AccountService {
    val apiKey = idexjaxrsExchange.exchangeSpecification.apiKey

    init {

        if (IdexjaxrsExchange.debugMe) {
//            apiClient = ApiClient()
//            IdexjaxrsExchange.setupDebug(apiClient)
        }
    }

    override fun getAccountInfo(): AccountInfo {
        val apiKey = idexjaxrsExchange.exchangeSpecification.apiKey
        val s = apiKey.slice(0.rangeTo(6)) + "…"
        val returnBalancesPost = completeBalances(CompleteBalancesReq().apply { address = apiKey })

        val entries = returnBalancesPost.entries
        val toList = entries.toList()
        val list = toList.map {
            val key = it.key!!
            val value = it.value!!
            val currency = Currency(key)
            Balance(currency, null, value.available, value.onOrders)
        }
        val accountInfo = AccountInfo(
                Wallet(s, list))
        return accountInfo
    }

    override fun requestDepositAddress(currency: Currency?, vararg args: String?): String =//the safe option
            idexjaxrsExchange.exchangeSpecification.apiKey


    override fun getFundingHistory(params: TradeHistoryParams): MutableList<FundingRecord> {
        when {
            params is IdexTradeHistoryParams -> {
                val returnDepositsWithdrawalsPost = fundingHistory(params as DepositsWithdrawalsReq)
                return listOf(
                        returnDepositsWithdrawalsPost.deposits.map {
                            FundingRecord(
                                    apiKey,
                                    Date(it.timestamp.toLong() * 1000),
                                    Currency(it.currency),
                                    it.amount.toBigDecimalOrNull() ?: BigDecimal.ZERO,
                                    it.transactionHash,
                                    it.depositNumber,
                                    FundingRecord.Type.DEPOSIT,
                                    FundingRecord.Status.resolveStatus(it.status),
                                    BigDecimal.ZERO,
                                    BigDecimal.ZERO,
                                    ""
                            )
                        },
                        returnDepositsWithdrawalsPost.withdrawals.map {
                            FundingRecord(
                                    apiKey,
                                    Date(it.timestamp.toLong() * 1000),
                                    Currency(it.currency),
                                    it.amount.toBigDecimalOrNull() ?: BigDecimal.ZERO,
                                    it.transactionHash,
                                    it.depositNumber,
                                    FundingRecord.Type.WITHDRAWAL,
                                    FundingRecord.Status.resolveStatus(it.status),
                                    BigDecimal.ZERO,
                                    BigDecimal.ZERO,
                                    ""
                            )
                        }
                ).flatten().sortedBy { it.date }.toMutableList()
            }

            else -> throw  ApiException(
                    "getFundingHistory requires " + IdexTradeHistoryParams::class.java.canonicalName)
        }
    }

    override fun createFundingHistoryParams(): TradeHistoryParams {
        return IdexTradeHistoryParams()
    }

    /**
     * from the docs: "Useful response upon withdrawal success is in the works, for now simply test that there is no error property in the result object to confirm your withdrawal has succeeded."
     */
    override fun withdrawFunds(w: WithdrawFundsParams): String {
        when {
            w is IdexWithdraw -> {
                val withdraw = withdraw(w)
                return IdexjaxrsExchange.gson.toJson(withdraw)

            }
            else -> throw ApiException(
                    "withdraw method requires " + IdexWithdraw::class.java.canonicalName)
        }
    }


    override fun withdrawFunds(currency: Currency, amount: BigDecimal, address: String): String = withdrawFunds(
            IdexWithdraw(apiKey,
                         amount.toString(),
                         currency!!.symbol,
                         idexjaxrsExchange.tradeService.idexServerNonce,
                         s = TODO("provide threadlocals? for S"),
                         v = TODO("provide threadlocals? for V"),
                         r = TODO("provide threadlocals? for R"))
    )
}